File: python/problems/pg42587/solution.py
"""
[PG] 42587 - 프로세스
https://programmers.co.kr/learn/courses/30/lessons/42587
난이도: lv2
태그: stack,queue

시간복잡도: O(n^2)
공간복잡도: O(n)
"""
from collections import deque

def solution(priorities: list[int], location: int):
    queue = deque(enumerate(priorities))
    priorities.sort(reverse=True)
    count = 0
    while queue:
        idx, p = queue.popleft()

        if p < priorities[count]:
            queue.append((idx, p))
        else:
            count+=1
            if idx == location:
                return count
    return count

File: java/src/main/java/problems/pg42587/Solution.java
package problems.pg42587;

import java.util.*;

/**
 * [PG] 42587 - 프로세스 https://programmers.co.kr/learn/courses/30/lessons/42587 난이도: lv2 태그:
 * stack,queue
 *
 * <p>시간복잡도: O(N^2) 공간복잡도: O(N)
 */
public class Solution {

  public int solution(int[] priorities, int location) {

    final Queue<Process> queue = new ArrayDeque<>();
    for (int i = 0; i < priorities.length; i++) {
      queue.offer(new Process(i, priorities[i]));
    }
    priorities =
        Arrays.stream(priorities)
            .boxed()
            .sorted(Collections.reverseOrder())
            .mapToInt(Integer::intValue)
            .toArray();

    int count = 0;
    while (!queue.isEmpty()) {
      final Process process = queue.poll();
      if (process.priority() < priorities[count]) {
        queue.offer(process);
      } else {
        count++;
        if (process.idx() == location) {
          return count;
        }
      }
    }

    return count;
  }

  private record Process(int idx, int priority) {}
}
